010: thread spec (#7)
ssbc/sips: 010.md

> # Thread Spec

Reviewer: ```suggestion
# Threads
```

Authors: Mix Irving mix@protozoa.nz

Date: 2023-03-27

License: CC0-1.0

> 

Reviewer: ```suggestion

## Abstract

```

The `post` message type is the foundational unit for talking with peers on
the original scuttlebutt network. 

`post` messages can be used standalone, but are most commonly collected as
sets of related messages and formed into a "thread".

During rendering, threads of posts are also further _decorated_ with `vote`
messages (for likes), and `about` messages (for displaying author details).

> 

Reviewer: ```suggestion
## Terminology

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://tools.ietf.org/html/rfc2119).

## Specification

```

## 1. `post` schema

> 

Reviewer: ```suggestion

The following schemas relate to the `content` field in SSB messages. If a field is commented with `// optional`, it means it MAY appear in the `content` field, otherwise it SHOULD appear.

```

  text: 'This was an innovation @pfraze came up with',
  mentions: [
    {
      link: '@hxGxqPrplLjRG2vtjQL87abX4QKqeLgCwQpS730nNwE=.ed25519',
      name: '@pfraze'
    }
  ]
}
```
The mentions are then used to map the `text` to this before rendering:
```markdown
This was an innovation [@pfraze](@hxGxqPrplLjRG2vtjQL87abX4QKqeLgCwQpS730nNwE=.ed25519)
came up with',
```

### 5.3. Blocked authors

If some message `B` is from a blocked author, we may still use its tangle data
for causal ordering, but not render that message in the thread. 
  
The message MUST NOT be displayed to by deault in the UI, but it is up to the
particular client whether this means
- not rendering it at all
- rendering an empty skeleton message saying "message from blocked author"
- rendering a skeleton saying "message from blocked author, do you want to
  display?"
  

### 5.4. Content warnings

If the post contains a `contentWarning` field, the UI SHOULD NOT display the
raw content and instead display a message which:
1. alerts that this message has a content warning
2. displays the String in `contentWarning`

If the user clicks "show content" (or similar) then the message should be
rendered as usual.

> 

Reviewer: ```suggestion
## Reference implementation

- [ssb-threads](https://github.com/ssbc/ssb-threads)

```

Author: Need to check with you that you actually do the ordering like this spec suggests
I've seen the linearizing done a couple ways since I've worked on it

I think doing a sort like if A is upstream (in previous chain) of B then -1, if B is upstream of A, then +1, otherwise sort by timestamp?

Reviewer: ssb-threads uses [ssb-sort](https://github.com/ssbc/ssb-sort/) (which is the reference implementation for the causal/topo sort) so I think it's correct? https://github.com/ssbc/ssb-threads/blob/ca06fcafabf1870bcb29ea24d899235249b7966f/src/index.ts#L279

Author: cool. Yeah I remember the logic now, I think it's what I said...
https://github.com/ssbc/ssb-sort/blob/master/index.js#L32-L49

> ## References:

Reviewer: ```suggestion
## Informative references
```

## Specification
## 1. `post` schema

The following schemas relate to the `content` field in SSB messages. If a field is commented with `// optional`, it means it MAY appear in the `content` field, otherwise it SHOULD appear.

### 1.1. Root message

A "root message" is the first message that starts a particular thread

```javascript
{
  type: 'post',
  text: String,
  channel: String,       // optional
  contentWarning: String // optional
  mentions: [Mention],   // optional
  recps: [Recp]          // optional
}
```

### 1.2. Response message

These messages are used to post responses into a thread. They are NOT
used to reply to specific messages, rather to the currently held context
of a thread (see Tangles below).

```javascript
{
  type: 'post',
  text: String,
  root: MsgId,
  branch: [MsgId],
  channel: Channel,      // optional
  contentWarning: String // optional
  mentions: [Mention],   // optional
  recps: [Recp]          // optional
}
```

NOTE: sometimes `branch` will be a `MsgId` instead of `[MsgId]`
### 1.2. Channel

`Channel` should be the a `String` which:

> - is NOT `#` prefixed
> - has no spaces
> 
> Parsing it .... expect to need to filter leading `#` and to need to treat
> different capitalisations the same (e.g. `NewZealand` vs `newzealand`).
> 
> **TODO**: more detail (see `ssb-ref` ?)

Author: ```suggestion
- SHOULD NOT contain
    - hash symbol `#`
    - punctuation `,."!?`
    - brackets `()[]`
    - spaces ` `, 
- SHOULD NOT be longer than 30 characters long

When parsing a channel, you MUST normalize the channel name by:
1. removing disallowed characters
2. truncating to 30 characters (if required)
3. lowering the case of all characters
    - i.e. `NewZealand` is to be treated the same as `newzealand`, `newZealand` etc.

See [ssb-ref](https://github.com/ssbc/ssb-ref/blob/main/index.js#L95-L104) for a reference implementation of `normalizeChannel`
```

# Threads

Authors: Mix Irving mix@protozoa.nz

Date: 2023-03-27

License: CC0-1.0

## Abstract

The `post` message type is the foundational unit for talking with peers on
the original scuttlebutt network. 

`post` messages can be used standalone, but are most commonly collected as
sets of related messages and formed into a "thread".

During rendering, threads of posts are also further _decorated_ with `vote`
messages (for likes), and `about` messages (for displaying author details).

> 

Reviewer: ```suggestion

This SIP defines the expected schema for `post` and `vote` messages, and 
how they form a directed acyclic graph known as "threads" in SSB.

```

## Terminology

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://tools.ietf.org/html/rfc2119).

## Specification

## 1. `post` schema

The following schemas relate to the `content` field in SSB messages. If a field is commented with `// optional`, it means it MAY appear in the `content` field, otherwise it SHOULD appear.

### 1.1. Root message

A "root message" is the first message that starts a particular thread

```javascript
{
  type: 'post',
  text: String,
  channel: String,       // optional

>   contentWarning: String // optional

Reviewer: ```suggestion
  contentWarning: String, // optional
```

>   mentions: [Mention],   // optional

Reviewer: ```suggestion
  mentions: Array<Mention>,   // optional
```

I'm suggesting a TypeScript syntax here which should be a lot less ambiguous. One problem in this document is that the syntax `[x]` is ambiguous.

In one case (this case) `[Mention]` means an array of zero, one, or many mentions. But in another case,

```
B => { root: A, previous: [A] }
```

The notation `[A]` meant "an array with only ONE item, `A`".

So let's use `Array<Mention>`

>   recps: [Recp]          // optional

Reviewer: ```suggestion
  recps: Array<Recp>          // optional
```

}
```

### 1.2. Response message

These messages are used to post responses into a thread. They are NOT
used to reply to specific messages, rather to the currently held context
of a thread (see Tangles below).

```javascript
{
  type: 'post',
  text: String,
  root: MsgId,

>   branch: [MsgId],

Reviewer: ```suggestion
  branch: Array<MsgId>,
```

  channel: Channel,      // optional
  contentWarning: String // optional

>   mentions: [Mention],   // optional

Reviewer: ```suggestion
  mentions: Array<Mention>,   // optional
```

>   recps: [Recp]          // optional

Reviewer: ```suggestion
  recps: Array<Recp>          // optional
```

}
```

> NOTE: sometimes `branch` will be a `MsgId` instead of `[MsgId]`

Reviewer: ```suggestion
NOTE: sometimes `branch` MAY be a `MsgId` instead of `Array<MsgId>`
```

    - punctuation `,."!?`
    - brackets `()[]`
    - spaces ` `, 
- SHOULD NOT be longer than 30 characters long

When parsing a channel, you MUST normalize the channel name by:
1. removing disallowed characters
2. truncating to 30 characters (if required)
3. lowering the case of all characters
    - i.e. `NewZealand` is to be treated the same as `newzealand`, `newZealand` etc.

See [ssb-ref](https://github.com/ssbc/ssb-ref/blob/main/index.js#L95-L104) for a reference implementation of `normalizeChannel`

### 1.3. Mention

May take one of the following forms:

```javascript 
{
  link: FeedId,
  name: String  // optional
}
```
```javascript 
{
  link: MsgId
}
```
```javascript 
{
  link: BlobId,
  size: Int,    // optional
  name: String  // optional
}
```

You will also find `Mention` in the wild which are just a plain string
```
FeedId | MsgId | BlobId
```

## 2. `vote` Schema

`vote` messages are used for responding to `post` messages. Sometimes they are
called "likes" but they are used for expressing a range of emotional responses.

```javascript
{
  type: "vote",
  vote: {
    link: MsgId
    expression: String,
    value: 1
  },
  root: MsgId,         // optional

>   branch: [MsgId]      // opptional

Reviewer: ```suggestion
  branch: Array<MsgId>      // optional
```

```
_Figure. showing the creation of a tangle, then tie-breaking between concurrent
messages `C`,`D`, and best-guess insertion of disconnected message `Y`_

For **disconnected messages** we try to place them in the correct position by
timestamp alone. It could be a good idea to indicate in UI that this message may
be in the wrong place and is missing context.

```mermaid
flowchart RL
A(A):::root;B(B);C(C);D(D);M(M);N(N);Y(Y):::disconnected
N-->Y-->M-->D-->C-->B-->A
classDef default fill: #e953da, stroke-width: 0, color: white
classDef root fill: #4c44cf, stroke-width: 0, color: white
classDef disconnected stroke:none,fill:#e95353,color:none;
```
_Figure. showing a linearised thread tangle, along with a "best guess" placement
of disconnected message `Y`._

### 4.2. Scuttlebutt Markdown

The `text` field of `post` messages is Scuttlebutt flavoured Markdown, which is
basic Markdown with a couple of important privacy enhancing features.

#### 4.2.1 Image rendering

DO NOT render images that link to the web. This leaks data, which allows
tracking of users.
```markdown
![unsafe image](https://upload.wikimedia.org/wikipedia/commons/1/11/Panopticon.jpg)
```

> Such images MUST instead be rendered as a link with an alert/ warning

Reviewer: ```suggestion
Such images SHOULD instead be rendered as a link with an alert/ warning
```

I agree with the intent, but "MUST" is too strong, and it wasn't even used for the `post` msg schema, and it would make Manyverse not SIP10 compliant, which is a bummer. (Yes, Manyverse should be doing what you specced here, PRs welcome)

(see 4.2.2).

ONLY render images that link to blobs that are hosted locally:
```markdown
![snails](&7AUsuVcxUzIJZPBBLqR3O2KvIiuECgdYq9lzEh0jcS8=.sha256)
```

#### 4.2.2 Link rendering

> When a link to the web is present, the interface MUST render an alert to users

Reviewer: ```suggestion
When a link to the web is present, the interface SHOULD render an alert to users
```

I don't actually think an alert is important. In Manyverse we use black underscore links for WWW links and blue underscore links for SSB links and that's already one way of signaling to users.

Author: It was too subtle for me to know! Good to know though. Maybe www should be RED
